File: in_memory_cache_service/src/controllers/note_controller.ts
import { Request, Response } from "express";
import NoteRepositoryFactory from "../domains/note/note_repository_factory";
import { StatusCodes } from "http-status-codes";
import NoteEntityFactory from "../domains/note/note_entity_factory";
import VerifyIfUserExistsAdapter from "../infrastructure/concerns/verify_if_user_exists_adapter";

class NoteController {
  public static async getAllNotes(
    request: Request,
    response: Response,
  ): Promise<void> {
    try {
      const userId = Number(request.params.user_id);

      const noteRepository = await NoteRepositoryFactory.getInstance();

      const userNotes = await noteRepository.getNotes(userId);

      response.json(userNotes);
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
    } catch (error) {
      response.status(StatusCodes.NOT_FOUND).send("");
    }
  }

  public static async getNote(
    request: Request,
    response: Response,
  ): Promise<void> {
    try {
      const noteId = Number(request.params.id);

      const noteRepository = await NoteRepositoryFactory.getInstance();

      const note = await noteRepository.getNote(noteId);

      response.json(note);
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
    } catch (error) {
      response.status(StatusCodes.NOT_FOUND).send("");
    }
  }

  public static async createNote(
    request: Request,
    response: Response,
  ): Promise<void> {
    try {
      const userId = Number(request.body.userId);

      await VerifyIfUserExistsAdapter.verifyIfUserExists(userId);
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
    } catch (error) {
      response.status(StatusCodes.NOT_FOUND).send("");

      return;
    }

    try {
      const noteRepository = await NoteRepositoryFactory.getInstance();

      const noteId = Number(request.body.id);
      const noteTitle = request.body.title;
      const noteBody = request.body.body;
      const noteCreatedAt = Number(request.body.createdAt);
      const noteUpdatedAt = Number(request.body.updatedAt);
      const noteUserId = Number(request.body.userId);

      const note = NoteEntityFactory.getInstance(
        noteId,
        noteTitle,
        noteBody,
        noteCreatedAt,
        noteUpdatedAt,
        noteUserId,
      );

      await noteRepository.createNote(note);

      response.status(StatusCodes.CREATED).send("");
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
    } catch (error) {
      response.status(StatusCodes.CONFLICT).send("");
    }
  }

  public static async updateNote(
    request: Request,
    response: Response,
  ): Promise<void> {
    try {
      const noteRepository = await NoteRepositoryFactory.getInstance();

      const noteId = Number(request.body.id);
      const noteTitle = request.body.title;
      const noteBody = request.body.body;
      const noteCreatedAt = Number(request.body.createdAt);
      const noteUpdatedAt = Number(request.body.updatedAt);
      const noteUserId = Number(request.body.userId);

      const note = NoteEntityFactory.getInstance(
        noteId,
        noteTitle,
        noteBody,
        noteCreatedAt,
        noteUpdatedAt,
        noteUserId,
      );

      await noteRepository.updateNote(note);

      response.status(StatusCodes.OK).send("");
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
    } catch (error) {
      response.status(StatusCodes.NOT_FOUND).send("");
    }
  }

  public static async deleteNote(
    request: Request,
    response: Response,
  ): Promise<void> {
    try {
      const noteId = Number(request.params.id);
      const userId = Number(request.params.user_id);

      const noteRepository = await NoteRepositoryFactory.getInstance();

      await noteRepository.deleteNote(noteId, userId);

      response.json(StatusCodes.NO_CONTENT);
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
    } catch (error) {
      response.status(StatusCodes.NOT_FOUND).send("");
    }
  }
}

export default NoteController;

File: data_source_service/app/controllers/note_controller.rb
class NoteController < ApplicationController
  def get_all_notes
    begin
      id = params[:id]

      VerifyIfUserExistsCommand.execute id

      all_user_notes = NoteUseCases.get_all_notes id

      render json: all_user_notes, status:200
    rescue StandardError
      render status: 404
    end
  end

  def get_note
    begin
      note_id = params[:note_id]

      note = NoteUseCases.get_note note_id

      render json: note, status: 200
    rescue StandardError
      render status: 404
    end
  end

  def create_note
    begin
      serialize_request_body = JSON.parse request.raw_post

      id = serialize_request_body[NoteConstants::NOTE_MAP_ID_KEY]
      title = serialize_request_body[NoteConstants::NOTE_MAP_TITLE_KEY]
      body = serialize_request_body[NoteConstants::NOTE_MAP_BODY_KEY]
      created_at = serialize_request_body[NoteConstants::NOTE_MAP_CREATED_AT_KEY]
      updated_at = serialize_request_body[NoteConstants::NOTE_MAP_UPDATED_AT_KEY]
      user_id = params[:id]
      
      VerifyIfUserExistsCommand.execute user_id
      
      NoteUseCases.create_note id, title, body, created_at, updated_at, user_id

      render status: 201
    rescue ActiveRecord::RecordNotUnique
      render status: 409

    rescue ActiveRecord::RecordNotFound
      render status: 404
    end
  end

  def update_note
    begin
      serialize_request_body = JSON.parse request.raw_post

      id = params[:note_id]
      title = serialize_request_body[NoteConstants::NOTE_MAP_TITLE_KEY]
      body = serialize_request_body[NoteConstants::NOTE_MAP_BODY_KEY]
      updated_at = serialize_request_body[NoteConstants::NOTE_MAP_UPDATED_AT_KEY]
      user_id = params[:id]

      NoteUseCases.update_note(title, body, updated_at, id)

      render status: 200
    rescue StandardError
      render status: 404
    end
  end

  def delete_note
    begin
      id = params[:note_id]

      NoteUseCases.delete_note id

      render status: 204
    rescue StandardError
      render status: 404
    end
  end
end
